### PR TITLE
[Feat: GraphQL sidebar] GraphQL History

### PR DESCRIPTION
**Pull request by oshhh**
**Changes:**

* View graphQL history
* Add, delete, clear history
* Use history

**UI**

image

**tasks:**
- [x] introduce history component in graphql page
- [x] separate history entry component for rest and graphql
- [x] expand / collapse graphql query on entry
- [x] sync with firebase and localstorage
- [x] tests

**checks:**
- [x] add / delete history entry
- [x] add / remove star from entry
- [x] clear all entries
- [x] sync with firebase and localstorage

**deploy preview:** https://deploy-preview-1528--hoppscotch.netlify.app

**remarks: PR contains 20+ commits: squash and merge is preferred.**

### CODE REVIEW

File: pages/index.vue [hunk 1]
```diff
@@ -532,7 +532,7 @@
         <section>
           <SmartTabs>
             <SmartTab :id="'history'" :label="$t('history')" :selected="true">
-              <HttpHistory @useHistory="handleUseHistory" ref="historyComponent" />
+              <History v-bind:page="'rest'" @useHistory="handleUseHistory" ref="historyComponent" />
```
Comment (liyasthomas) [inline pages/index.vue:535]:
```suggestion
              <History :page="'rest'" @useHistory="handleUseHistory" ref="historyComponent" />
```
Let's stick with using shorthands for binding props

File: pages/graphql.vue [hunk 2]
```diff
@@ -696,6 +738,20 @@ export default {
         this.$toast.info(this.$t("finished_in", { duration }), {
           icon: "done",
         })
+
+        entry = {
+          ...entry,
+          response: this.response,
+          date: new Date().toLocaleDateString(),
+          time: new Date().toLocaleTimeString(),
```
Comment (liyasthomas) [inline pages/graphql.vue:746]:
```suggestion
```
No need for separate date and time attributes.

File: pages/graphql.vue [hunk 3]
```diff
@@ -696,6 +738,20 @@ export default {
         this.$toast.info(this.$t("finished_in", { duration }), {
           icon: "done",
         })
+
+        entry = {
+          ...entry,
+          response: this.response,
+          date: new Date().toLocaleDateString(),
```
Comment (liyasthomas) [inline pages/graphql.vue:745]:
```suggestion
          updatedOn: new Date(),
```
We generally use `updatedOn` attribute to store timestamps. Also, saving timestamp as string isn't preferred. Firebase support date datatype. Must leverage that.

File: pages/graphql.vue [hunk 4]
```diff
@@ -528,6 +550,18 @@ export default {
     next()
   },
   methods: {
+    useSelectedEnvironment(event) {
+      console.log("use selected environment")
+    },
+    handleUseHistory(entry) {
+      console.log("use history")
```
Comment (liyasthomas) [inline pages/graphql.vue:557]:
```suggestion
```
Remove console logs.

File: components/history/index.vue [hunk 5]
```diff
@@ -218,56 +114,78 @@
...
+    name: "getUser",
+    method: "query",
+    query: `query getUser($uid: String!) {
+  user(uid: $uid) {
+    name
+  }
+}`,
+    star: false,
+  },
+]
```
Comment (liyasthomas) [inline components/history/index.vue:175]:
```suggestion
```

Delete unused variable

**Review by liyasthomas — CHANGES_REQUESTED** [review]

File: components/history/graphql/card.vue [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+          </div>
+        </template>
+      </v-popover>
+    </div>
+    <div class="show-on-large-screen">
+      <li>
+        <input
+          v-for="(line, index) in expand
+            ? entry.query.split('\n')
+            : entry.query.split('\n').slice(0, 2).concat(['...'])"
```
Comment (liyasthomas) [inline components/history/graphql/card.vue:72]: You can't use expand (`boolean` data type) on `v-for`. `ternary operator` doesn't seems to be working inline inside `v-for`. To iterate over `entry.query`, you can go with an approach similar to this:

```
          v-for="(line, index) in getExpandQueryLines"
```

```
export default {
	  ...
	  ...
	computed: {
		getExpandQueryLines() {
			return this.expand ? entry.query.split('\n') : entry.query.split('\n').slice(0, 2).concat(['...'])
		}
	}
	  ...
	  ...
}
```

Haven't tested this code, please let me know if this fix DeepCode analysis. @oshhh

File: pages/graphql.vue (hunk 3 above)
Comment (oshhh) [reply pages/graphql.vue:743]: So I've tried to follow the format used in index.vue for storing REST requests. Should I change it there too?

Comment (liyasthomas) [reply pages/graphql.vue:742]: That would be good.

File: pages/graphql.vue [hunk 7]
```diff
@@ -318,79 +318,102 @@
...
-              <div class="gqlTabs">
-                <SmartTab
-                  v-if="queryFields.length > 0"
-                  :id="'queries'"
-                  :label="$t('queries')"
-                  :selected="true"
-                >
-                  <div v-for="field in filteredQueryFields" :key="field.name">
-                    <GraphqlField :gqlField="field" :jumpTypeCallback="handleJumpToType" />
+        <SmartTabs styles="m-4">
```
Comment (liyasthomas) [inline pages/graphql.vue:321]:
```suggestion
        <SmartTabs>
```
Doesn't require margin here.

File: components/history/graphql/card.vue [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+            </button>
+          </div>
+        </template>
+      </v-popover>
+    </div>
+    <div class="show-on-large-screen">
+      <li>
+        <input
+          v-for="(line, index) in query"
+          v-bind:key="index"
```
Comment (liyasthomas) [inline components/history/graphql/card.vue:71]:
```suggestion
          :key="`line-${index}`"
```
use shortcut for v-bind

File: components/history/index.vue [hunk 9]
```diff
@@ -13,127 +13,23 @@
...
+          v-bind:entry="entry"
+          v-bind:showMore="showMore"
+          @toggle-star="toggleStar(entry)"
+          @delete-entry="deleteHistory(entry)"
+          @use-entry="useHistory(entry)"
+        />
+        <HistoryGraphqlCard
+          v-if="page == 'graphql'"
+          v-bind:entry="entry"
+          v-bind:showMore="showMore"
```
Comment (liyasthomas) [inline components/history/index.vue:28]:
```suggestion
          :entry="entry"
          :showMore="showMore"
          @toggle-star="toggleStar(entry)"
          @delete-entry="deleteHistory(entry)"
          @use-entry="useHistory(entry)"
        />
        <HistoryGraphqlCard
          v-if="page == 'graphql'"
          :entry="entry"
          :showMore="showMore"
```
use shortcut for v-bind

File: pages/graphql.vue [hunk 10]
```diff
@@ -318,79 +318,102 @@
...
+                {{ $t("send_request_first") }}
+              </p>
+            </AppSection>
+          </SmartTab>
+
+          <SmartTab :id="'history'" :label="$t('history')">
+            <History
+              @useHistory="handleUseHistory"
+              ref="graphqlHistoryComponent"
+              v-bind:page="'graphql'"
```
Comment (liyasthomas) [inline pages/graphql.vue:401]:
```suggestion
              :page="'graphql'"
```
use shortcut for v-bind

File: components/history/graphql/card.vue [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+      </button>
+      <button
+        class="icon"
+        @click="expand = !expand"
+        v-tooltip="{
+          content: expand ? 'show less' : 'show more',
+        }"
+      >
+        <i class="material-icons">
+          {{ expand ? "compress" : "expand" }}
```
Comment (liyasthomas) [inline components/history/graphql/card.vue:34]:
```suggestion
          {{ expand ? "unfold_less" : "unfold_more" }}
```
use unfold icons

File: components/history/index.vue [hunk 12]
```diff
@@ -13,127 +13,23 @@
...
-                type="text"
-                readonly
-                :value="entry.preRequestScript"
-                :placeholder="$t('no_prerequest_script')"
-                class="pt-0 mt-0 text-sm bg-transparent text-fgLightColor"
-              />
-            </li>
-          </div>
-        </transition>
+      <ul v-for="(entry, index) in filteredHistory" :key="'card_' + index">
```
Comment (liyasthomas) [inline components/history/index.vue:16]:
```suggestion
      <ul v-for="(entry, index) in filteredHistory" :key="`entry-${index}`">
```
use template literals in :key

**Review by liyasthomas — CHANGES_REQUESTED** [review]

File: components/history/graphql/card.vue [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+      >
+        <i class="material-icons">
+          {{ entry.star ? "star" : "star_border" }}
+        </i>
+      </button>
+      <button
+        class="icon"
+        @click="expand = !expand"
+        v-tooltip="{
+          content: expand ? 'show less' : 'show more',
```
Comment (liyasthomas) [inline components/history/graphql/card.vue:30]:
```suggestion
          content: expand ? $t('hide_more') : $t('show_more'),
```
minor translation

File: components/history/index.vue [hunk 14]
```diff
@@ -298,43 +195,37 @@ export default {
...
-        }
-      )
-    },
     async deleteHistory(entry) {
-      if (fb.currentUser !== null) {
-        await fb.deleteHistory(entry)
-      }
+      console.log(this.history)
       this.history.splice(this.history.indexOf(entry), 1)
+      console.log(this.history)
```
Comment (liyasthomas) [inline components/history/index.vue:214]:
```suggestion
      this.history.splice(this.history.indexOf(entry), 1)
```
remove console logs

File: components/history/graphql/card.vue [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+@media (max-width: 720px) {
+  .virtual-list.filled {
+    min-height: 320px;
+  }
+
+  .labels {
+    display: none;
+  }
+}
```
Comment (liyasthomas) [inline components/history/graphql/card.vue:153]:
```suggestion
.stared {
  color: #f8e81c !important;
}
```
remove unused css

File: components/history/rest/card.vue [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+            :value="entry.preRequestScript"
+            :placeholder="$t('no_prerequest_script')"
+            class="pt-0 mt-0 text-sm bg-transparent text-fgLightColor"
+          />
+        </li>
+      </div>
+    </transition>
+  </div>
+</template>
+
```
Comment (liyasthomas) [inline components/history/rest/card.vue:125]:
```suggestion

<style scoped lang="scss">
.stared {
	color: #f8e81c !important;
}
</style>

```

File: components/history/index.vue [hunk 17]
```diff
@@ -218,56 +114,53 @@
...
 }
-
 .fade-enter,
 .fade-leave-to {
   opacity: 0;
 }
-
 .stared {
   color: #f8e81c !important;
 }
```
Comment (liyasthomas) [inline components/history/index.vue:131]:
```suggestion
```
remove absolute styles

**Review by liyasthomas — APPROVED** [review]
LGTM

**Comment by AndrewBastin** [comment]
Looks good to me!

But make sure to write tests for the new components.

After that I will give the approval. ^_^

**Review by AndrewBastin — CHANGES_REQUESTED** [review]
Looks good to me!

But make sure to write tests for the new components.

After that I will give the approval. ^_^

**Review by AndrewBastin — CHANGES_REQUESTED** [review]
Please check the marked comments.

File: pages/index.vue [hunk 18]
```diff
@@ -1363,8 +1363,7 @@ export default {
           const entry = {
             name: this.requestName,
             status: this.response.status,
-            date: new Date().toLocaleDateString(),
-            time: new Date().toLocaleTimeString(),
```
[the remark this answers is not in the document]
Comment (oshhh) [reply pages/index.vue:1367]: Okay, should I revert this change?

**Review by AndrewBastin — APPROVED** [review]
LGTM 💯

**Pull request merged**